### PR TITLE
don't erase hostconfig on refresh

### PR DESCRIPTION
**Pull request by vieux**
Fix: https://github.com/docker/swarm/issues/1846

### CODE REVIEW

**Comment by vieux** [comment]
ping @nishanttotla

File: cluster/engine.go [hunk 1]
```diff
@@ -620,6 +620,9 @@ func (e *Engine) updateContainer(c dockerclient.Container, containers map[string
 		}
 		// Convert the ContainerConfig from inspect into our own
 		// cluster.ContainerConfig.
+		if info.HostConfig != nil {
+			info.Config.HostConfig = *info.HostConfig
```
Comment (nishanttotla) [inline cluster/engine.go]: why is the explicit dereference needed here?

File: cluster/engine.go [hunk 2]
```diff
@@ -620,6 +620,9 @@ func (e *Engine) updateContainer(c dockerclient.Container, containers map[string
...
+		}
 		container.Config = BuildContainerConfig(*info.Config)
 
 		// FIXME remove "duplicate" lines and move this to cluster/config.go
```
Comment (nishanttotla) [inline cluster/engine.go]: In the lines below this, are the HostConfig updates needed anymore?

File: cluster/engine.go (hunk 1 above)
Comment (vieux) [inline cluster/engine.go]: because `HostConfig` embed in `Config` is an object, not a pointer

File: cluster/engine.go (hunk 2 above)
Comment (vieux) [inline cluster/engine.go]: yes

File: cluster/engine.go (hunk 1 above)
Comment (nishanttotla) [inline cluster/engine.go]: Right, my question was about the RHS. I didn't realize the `HostConfig` inside info is a pointer, so my bad.

**Comment by nishanttotla** [comment]
LGTM.

**Comment by nishanttotla** [comment]
Also ping @abronan seems like `21 leader election` test fails often.

**Comment by dongluochen** [comment]
LGTM

**Comment by nishanttotla** [comment]
Ping @vieux I think we should merge this.

**Comment by abronan** [comment]
LGTM

**Pull request merged**